mesonbuild/modules/python.py

```diff
@@ -276,11 +276,57 @@ def set_env(name, value):
 
 
 INTROSPECT_COMMAND = '''\
+import os.path
 import sysconfig
 import json
 import sys
 
-install_paths = sysconfig.get_paths(vars={'base': '', 'platbase': '', 'installed_base': ''})
+def get_distutils_paths(*, scheme=None, prefix=None):
+    import distutils.dist
+    distribution = distutils.dist.Distribution()
+    install_cmd = distribution.get_command_obj('install')
+    if prefix is not None:
+        install_cmd.prefix = prefix
+    if scheme:
+        install_cmd.select_scheme(scheme)
+    install_cmd.finalize_options()
+    return {
+        'data': install_cmd.install_data,
+        'include': os.path.dirname(install_cmd.install_headers),
+        'platlib': install_cmd.install_platlib,
+        'purelib': install_cmd.install_purelib,
+        'scripts': install_cmd.install_scripts,
+    }
+
+# On Debian derivatives, the Python interpreter shipped by the distribution uses
+# a custom install scheme, deb_system, for the system install, and changes the
+# default scheme to a custom one pointing to /usr/local and replacing
+# site-packages with dist-packages.
+# See https://github.com/mesonbuild/meson/issues/8739.
+# XXX: We should be using sysconfig, but Debian only patches distutils.
+
+try:
+    # Debian patches distutils, keeping only distutils.version, so
+    # `import distutils` will always work, but
+    # `import distutils.command.install` will fail if python3-distutils is
+    # not installed. Let's try detecting that and error out with an helpful
+    # error message.
+    import distutils
+    try:
+        import distutils.command.install
+    except ModuleNotFoundError:
+        raise ModuleNotFoundError('Unable to import distutils, please install python3-distutils')
+    debian_distutils = 'deb_system' in distutils.command.install.INSTALL_SCHEMES
+except ModuleNotFoundError:
+    debian_distutils = False
```

Comment:
I don't think we need a try inside a try, `import distutils` should always work on all platforms, right?

Reply:
Not after 3.12.

Reply:
Unless someone is naughty enough to create a Python package that provides the same module.

Reply:
Unless someone is naughty enough to create a Python package that provides the same module.

Reply:
> In that case we'll need to fix links_against_libpython() too, let's cross that bridge when/if we get there? OTOH, this is getting petty nitpicking I admit ;P

I'd prefer doing in a different PR as it is not actually an issue right now, while this is. We can fix the current handling of the Debian install paths and then explore how to fase out distutils entirely.

> OOC, how confident are we that it's going to be removed in 3.12?

Very.

> It has been formally announced already? When that happens, what's the likely outcome for Debian, they'll patch sysconfig instead?

Yes, my best guess is either PEP 668 or that, or both.
Also related: https://bugs.python.org/issue43976 (maybe skip down to the latest discussion)

> Sorry again for all my questions, I wasn't familiar with this topic, so I like to learned it a bit :)

No worries, I have been dealing with these issues for a little while and still struggle to keep track of everything.